File: Day_2/threeSum.py
def threeSum(nums):
    # Brute Force Method (Using 3 nested loops)
    # ans = []
    # for i in range(len(nums)):
    #     for j in range(i + 1, len(nums)):
    #         for k in range(j + 1, len(nums)):
    #             if nums[i] + nums[j] + nums[k] == 0:
    #                 triplet = sorted([nums[i], nums[j], nums[k]])
    #                 if triplet not in ans:
    #                     ans.append(triplet)
    # return ans
    
    

    # Better Solution (Using Hash Map)
    # ans = []
    # if len(nums) < 3:
    #     return ans
    # for i in range(len(nums)):
    #     mpp = {}
    #     for j in range(i + 1, len(nums)):
    #         target = -1 * (nums[i] + nums[j])
    #         if target in mpp:
    #             triplet = sorted([nums[i], nums[j], target])
    #             if triplet not in ans:
    #                 ans.append(triplet)
    #         mpp[nums[j]] = 1
    # return ans
    
    
    

    # Optimal Solution (Sorting + Two Pointer Approach)
    ans = []
    nums.sort()

    for i in range(len(nums)):
        if i != 0 and nums[i] == nums[i - 1]:
            continue
        j = i + 1
        k = len(nums) - 1
        while j < k:
            sum_ = nums[i] + nums[j] + nums[k]
            if sum_ == 0:
                ans.append([nums[i], nums[j], nums[k]])
                j += 1
                k -= 1
                while j < k and nums[j] == nums[j - 1]:
                    j += 1
                while j < k and nums[k] == nums[k + 1]:
                    k -= 1
            elif sum_ < 0:
                j += 1
            else:
                k -= 1

    return ans



def main():
        


    question = [-1, 0, 1, 2, -1, -4]
    ans = threeSum(question)

    
    print(ans)


if __name__ == "__main__":
    main()

File: Day_2/threeSum.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> threeSum(vector<int> nums){
    // // Brute Force Method

        // set<vector<int>> set;
        // if(nums.size() < 3 ) return {};
        // for(int i = 0 ; i < nums.size() ; i++){
        //     for(int j = i + 1 ; j < nums.size() ; j++){
        //         for(int k = j + 1 ; k < nums.size() ; k++){
        //             if(nums[i] + nums[j] + nums[k] == 0){
        //                 vector<int> a = {nums[i], nums[j], nums[k]};
        //                 sort(a.begin(), a.end());
        //                 set.insert(a);
        //             }
        //         }
        //     }
        // }

        // vector<vector<int>> ans;

        // for(auto i : set){
        //     ans.push_back(i);
        // }
        // return ans;


        
        //// Hash Map Solution
        
        // set<vector<int>> res;
        // if(nums.size() < 3) return {};
        // for (int i = 0 ; i < nums.size() ; i++){
        //     map<int, int> mpp;
        //     // vector<int> a;
        //     // int sum = nums[i];
        //     for(int j = i + 1 ; j < nums.size() ; j++){
        //         int sum = nums[i] + nums[j];
        //         int target = -1 * sum;
        //         if(mpp.find(target) != mpp.end()){
        //             vector<int> triplet = {nums[i], nums[j], target};
        //             sort(triplet.begin(), triplet.end()); 
        //             res.insert(triplet);
        //         }
        //         mpp[nums[j]]++;
        //     }
        // }
        

        // vector<vector<int>> ans;
        // for (auto i : res){
        //     ans.push_back(i);
        // }

        // return ans;


        // // OPTIMAL SOLUTION

        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());

        for(int i = 0 ; i < nums.size() ; i++){
            if(i != 0 && nums[i] == nums[i - 1]) continue;
            int j = i + 1;
            int k = nums.size() - 1;
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0){
                    vector<int> a = {nums[i], nums[j], nums[k]};
                    ans.push_back(a);
                    j++;
                    k--;
                    while(j < k && nums[j] == nums[j - 1]) {j++;}
                    while(j < k && nums[k] == nums[k + 1]) {k--;}
                }else if(sum < 0){
                    j++;
                } else{
                    k--;
                }
            }
            
        }
    
        return ans;
}



int main(){

    vector<int> question = {-1,0,1,2,-1,-4};
    vector<vector<int>> ans;
    ans = threeSum(question);

    for (vector<int> i : ans){
        for(int j : i){
            cout << j << " ";
        }cout << endl;
    }

}